Adding General Development Guidelines (#252)
cerner/terra-ui: src/terra-dev-site/guide/GeneralGuidelines.l.guide.mdx

# General Guidelines

> Keep in mind these things when developing terra components

Reviewer: ```suggestion
Keep these things in mind when developing with Terra components:
```

Reviewer 2: ![](https://media.giphy.com/media/3ohuAxV0DfcLTxVh6w/giphy.gif)

> * Terra components are black boxes. You should not rely on internal implementation specifics. This includes DOM structure, classes, and undocumented side effects.

Reviewer 2: Should we mention that css files are internal implementation and not considered part of the api?

> * Terra components are black boxes. You should not rely on internal implementation specifics. This includes DOM structure, CSS specifics, themeable variables, class names, and undocumented side effects.

Reviewer 3: ```suggestion
* Terra components are black boxes. You should not rely on internal implementation specifics. This includes DOM structure, CSS specifics, themeable variables, class names, and undocumented side effects. These details are not part of the documented API contract and are subject to change at any time.
```

> * Try and keep your DOMs small. If you have a list that could have thousands of items in it, you might need to rethink your design

Reviewer: ```suggestion
* Try to minimize your application's DOM footprint. An application's DOM size is inversely proportional to its performance. If you have a list that could have thousands of items in it, you need to rethink your design.
```

> * Take some time to learn the basics of JavaScript and React

Reviewer: Links to resources might be helpful

Reviewer 2: https://developer.mozilla.org/en-US/docs/Web/JavaScript

> # General Guidelines

Reviewer 4: Can we move this to come 2nd in the list
```
What is Terra
General Guidelines
How to Install
Webpack etc

```

Keep these things in mind when developing with Terra components:

* Terra components are black boxes. You should not rely on internal implementation specifics. This includes DOM structure, CSS specifics, themeable variables, class names, and undocumented side effects. These details are not part of the documented API contract and are subject to change at any time.
* Try to minimize your application's DOM footprint. An application's DOM size is inversely proportional to its performance. If you have a list that could have thousands of items in it, you need to rethink your design.

> * Take some time to learn the basics of [JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript) and [React](https://reactjs.org/tutorial/tutorial.html)

Reviewer 2: ```suggestion
* Take some time to learn the basics of [JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript) and [React](https://reactjs.org/tutorial/tutorial.html).
```

> * Pay attention to your dependencies. Dependencies that are only used in development should be dev dependencies. Make sure all peer dependencies are met.

Reviewer 4: should we add be sure learn about NPM dependency management & link to NPM

> * Use Terra's [webpack](https://engineering.cerner.com/terra-ui/getting-started/terra-ui/webpack) configuration where possible

Reviewer 2: ```suggestion
* Use Terra's [webpack](https://engineering.cerner.com/terra-ui/getting-started/terra-ui/webpack) configuration where possible.
```
